File: code/python/ch04/practice/practice_01_calculator.py
"""
Practice 1: Calculator
==============================
Chapter 4: Functions

PROBLEM
-------
Build a mini calculator. Define separate helper functions for add,
subtract, multiply, and divide. The solve function dispatches to the
correct helper based on the operator string.

INPUT FORMAT
------------
Three values on separate lines: int a, string op, int b.
op is one of: "+", "-", "*", "/"

OUTPUT FORMAT
-------------
The integer result, or "None" for division by zero or invalid operator.

CONSTRAINTS
-----------
- a, b are integers
- op is a single-character string
- For division, use integer division (//)
- Division by zero returns None
- Invalid operator returns None

EXAMPLES
--------
Input:  10, +, 3
Output: 13

Input:  10, /, 0
Output: None

Input:  10, ^, 3
Output: None

INSTRUCTIONS
------------
Replace the `pass` in each helper function and solve() with your solution.
The main block at the bottom handles input/output — don't change it.
"""


def add(a: int, b: int) -> int:
    """Return a + b."""
    pass  # TODO: Replace this with your solution


def subtract(a: int, b: int) -> int:
    """Return a - b."""
    pass  # TODO: Replace this with your solution


def multiply(a: int, b: int) -> int:
    """Return a * b."""
    pass  # TODO: Replace this with your solution


def divide(a: int, b: int) -> int | None:
    """Return a // b, or None if b is zero."""
    pass  # TODO: Replace this with your solution


def solve(a: int, op: str, b: int) -> int | None:
    """Dispatch to the correct operation based on op."""
    pass  # TODO: Replace this with your solution


# ── Do not change anything below this line ──────────────────────────
if __name__ == "__main__":
    a = int(input())
    op = input().strip()
    b = int(input())
    print(solve(a, op, b))

File: code/java/ch04/practice/Practice01Calculator.java
package ch04.practice;

import java.util.Scanner;

/**
 * Practice 01: Calculator
 * ==============================
 * Chapter 4: Functions
 *
 * PROBLEM
 * -------
 * Build a simple calculator using helper functions. Write separate
 * helper methods for add, subtract, multiply, and divide. Then write
 * a solve() method that picks the right operation based on the operator
 * string.
 *
 * INPUT FORMAT
 * ------------
 * Three values on one line: int operator int
 * Example: 10 + 3
 *
 * OUTPUT FORMAT
 * -------------
 * Print the result as an integer.
 * Print "Error" for invalid operator or division by zero.
 *
 * CONSTRAINTS
 * -----------
 * - a, b are integers
 * - op is one of: "+", "-", "*", "/"
 * - Division is integer division (truncated toward zero)
 *
 * EXAMPLES
 * --------
 * Input:  10 + 3
 * Output: 13
 *
 * Input:  10 / 3
 * Output: 3
 *
 * Input:  10 / 0
 * Output: Error
 *
 * Input:  10 % 3
 * Output: Error
 *
 * INSTRUCTIONS
 * ------------
 * 1. Implement add(), subtract(), multiply(), divide() helpers.
 * 2. Implement solve() using those helpers.
 * 3. Return null for invalid operator or divide-by-zero.
 * The main method handles input/output -- don't change it.
 */
public class Practice01Calculator {

    /** Add two numbers. */
    public static int add(int a, int b) {
        // TODO: Implement
        return 0;
    }

    /** Subtract b from a. */
    public static int subtract(int a, int b) {
        // TODO: Implement
        return 0;
    }

    /** Multiply two numbers. */
    public static int multiply(int a, int b) {
        // TODO: Implement
        return 0;
    }

    /** Divide a by b (integer division). Returns null if b is 0. */
    public static Integer divide(int a, int b) {
        // TODO: Implement (return null if b == 0)
        return null;
    }

    /**
     * Perform the operation and return the result.
     *
     * @param a  first operand
     * @param op the operator: "+", "-", "*", "/"
     * @param b  second operand
     * @return the result, or null for invalid op / divide-by-zero
     */
    public static Integer solve(int a, String op, int b) {
        // TODO: Replace this with your solution
        return null;
    }

    // ── Do not change anything below this line ──────────────────────
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int a = sc.nextInt();
        String op = sc.next();
        int b = sc.nextInt();
        Integer result = solve(a, op, b);
        if (result == null) {
            System.out.println("Error");
        } else {
            System.out.println(result);
        }
        sc.close();
    }
}
